pkg/networkdevice/diagnoses/diagnoses.go

```diff
@@ -0,0 +1,82 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2023-present Datadog, Inc.
+
+// Package diagnoses implements the diagnosis collection for NDM resources
+package diagnoses
+
+import (
+	"fmt"
+	"github.com/DataDog/datadog-agent/pkg/diagnose/diagnosis"
+	"github.com/DataDog/datadog-agent/pkg/networkdevice/metadata"
+)
+
+// Diagnoses hold diagnoses for a NDM resource
+type Diagnoses struct {
+	resourceType  string
+	resourceID    string
+	diagnoses     []metadata.Diagnosis
+	prevDiagnoses []metadata.Diagnosis
+}
+
+var severityMap = map[string]diagnosis.Result{
+	"success": diagnosis.DiagnosisSuccess,
+	"error":   diagnosis.DiagnosisFail,
+	"warn":    diagnosis.DiagnosisWarning,
+}
+
+// NewDeviceDiagnoses returns a new Diagnoses for a NDM device resource
+func NewDeviceDiagnoses(deviceID string) *Diagnoses {
+	return &Diagnoses{
+		resourceType: "ndm_device",
+		resourceID:   deviceID,
+	}
+}
+
+// Add adds a diagnoses
+func (d *Diagnoses) Add(result string, errorCode string, message string) {
+	d.diagnoses = append(d.diagnoses, metadata.Diagnosis{
+		Severity:  result,
+		ErrorCode: errorCode,
+		Diagnosis: message,
+	})
+}
+
+// ReportDiagnosis returns diagnosis metadata
+func (d *Diagnoses) ReportDiagnosis() []metadata.DiagnosisMetadata {
+	if len(d.prevDiagnoses) > 0 && len(d.diagnoses) == 0 {
```

Comment:
I'm wondering if we can avoid having the additional "previous diagnoses" by letting the caller `devicecheck.go` be responsible for the reset. By:

- Only tracking `diagnoses` list to track (without tracking previous diagnoses)
- ConvertToCLI always report `diagnoses` 
- When `devicecheck.go`  https://github.com/DataDog/datadog-agent/blob/1146bc6fb57c06f4ffcfc9283ebc5b31f27cc5b5/pkg/collector/corechecks/snmp/internal/devicecheck/devicecheck.go#L152 calls `ReportDiagnosis`, it will immediately also call `diagnoses.Reset()`.

WDYT?

Reply:
The reason prevDiagnoses is used is to avoid sending payloads with empty diagnoses

In this first iteration, I expect most of the devices to not emit diagnoses. To avoid send empty arrays each time for these, the following solution is used :

- if a device had no diagnoses during the last check run and the previous one, we don't send a diagnoses payload at all
- However if a device had diagnoses during the previous check run but not in the last one, we have to send an empty array of diagnoses so that backend can be cleared (hence the condition [here](https://github.com/DataDog/datadog-agent/pull/18973/files/b2d9586c7908c82c3217044f362c69d3e6378b53#diff-592a04efc53cc1bb1ec5344b7aaf7eadf5a7d96cffcc7d53b71380dff5abe58eR48))

We can drop the usage of prevDiagnosis but this will lead to having an empty diagnoses payload for each device, let me know what you think

Reply:
Indeed, thanks for explaining, that's an interesting optimisation.

That said, that's possibly a case of premature optimisation, by this I mean, sending diagnoses all the time (even empty) shouldn't cause perf issues:
- the payloads have similar volumes as devices, which is quite low compared to other things like interfaces.
- the size of the payload should be quite small
- adds some complexity

Also, (possibly unlikely) adding this logic to prevent sending empty diagnoses can possibly lead to some sneaky bugs. 

For example:
- check run 1: some diagnoses are send
- check run 2: empty diagnoses
- check run 3: empty diagnoses (not payload sent)
- check run 4: empty diagnoses (not payload sent)

If the payload at check run 2 is lost (rare case) for some reason, the diagnoses in the backend won't be cleaned and will still have diagnoses from check run 1.

Reply:
Good points, if the payloads are not a concern then let's remove this logic

Small nuance however : `diagnoses.Reset()` has to be called at the start of the check and not directly after `diagnoses.Report()`, or else no diagnoses will be collected when running `agent diagnose` (except if run during check execution)

Just updated the code with this and removed `prevDiagnoses` 🙂